### PR TITLE
journal-remote: Use monotonic time of receiver machine in log timestamp

### PR DESCRIPTION
**Pull request by chaloulo**
Replace the monotonic timestamp of the log entry with the monotonic timestamp of the receiver machine. This is useful when the receiver receives logs from multiple machines. In this case it is not guaranteed that the monotonic timestamps of the arriving log entries will be increasing, as they come from multiple machines. This causes the journal file to continously rotate, filling up the disk space very quickly.
In order to preserve all the information in the original log, the field _SOURCE_MONOTONIC_TIMESTAMP is appended to the log entry, with its value being the monotonic timestamp of the sender machine. For example, the following line is appended:
_SOURCE_MONOTONIC_TIMESTAMP=512142575

This pull request is a child of a previous pull request, https://github.com/systemd/systemd/pull/1531.
The code here represents one part of the functionality in the original pull request, also taking intoi account review comments.
The original pull request will be closed.

### CODE REVIEW

**Comment by evverx** [comment]
Are there any [doc](http://www.freedesktop.org/software/systemd/man/systemd.journal-fields.html) updates?

**Comment by chaloulo** [comment]
Hello,

Added documentation in man/systemd.journal-fields.xml. Also corrected an important typo at src/journal-remote/journal-remote-parse.c:450.

Best regards,

Klearchos

**Comment by evverx** [comment]
@chaloulo , thanks!

> The original monotonic timestamp of the log entry. systemd-journal-remote stores the monotonic timestamp of the sender machine here.

Hm.
I send a log entry `E` from `A` to `B`. Then, I send `E` from `B` to `C`. Looks like, I lose the original timestamp.

**Comment by chaloulo** [comment]
OK, I'll try to make the addition of the _SOURCE_MONOTONIC_TIMESTAMP conditional then. Only if no _SOURCE_MONOTONIC_TIMESTAMP exists, it will be created.

**Comment by chaloulo** [comment]
Hello,

I have made the addition of the _SOURCE_MONOTONIC_TIMESTAMP conditional. Only if it does not exist it will be added. So if the journal-remote logs are forwarded to a different machine, the _SOURCE_MONOTONIC_TIMESTAMP will still show the original monotonic timestamp of the log entry.

Best regards,

Klearchos

File: man/systemd.journal-fields.xml [hunk 1]
```diff
@@ -321,6 +331,7 @@
           </variablelist>
         </listitem>
       </varlistentry>
+
```
Comment (zonque) [inline man/systemd.journal-fields.xml]: Spurious new line?

File: src/journal-remote/journal-remote-parse.c [hunk 2]
```diff
@@ -441,9 +441,47 @@ static int process_data(RemoteSource *source) {
...
+    {
+        if (source->iovw.iovec[i].iov_len >= tag_len)
+        {
+            ret = memcmp(source->iovw.iovec[i].iov_base, "_SOURCE_MONOTONIC_TIMESTAMP=", tag_len);
+            if (ret==0)
+                break;
+        }
+    }
+    return ret==0 ? 1 : 0;
+}
```
Comment (zonque) [inline src/journal-remote/journal-remote-parse.c:460]: Please keep to the coding style implemented by other files in the source tree which is also described in `CODING_STYLE`. In particular, `{` are not put on a new line, we use spaces around `==`, before `*` in casts and after `;` in  `for` etc.

File: src/journal-remote/journal-remote-parse.c [hunk 3]
```diff
@@ -441,9 +441,47 @@ static int process_data(RemoteSource *source) {
         }
 }
 
+/*
+ * Returns 1 if field _SOURCE_MONOTONIC_TIMESTAMP already exists in the log entry,
+ * 0 if it does not.*/
+int source_monotonic_timestamp_exists(RemoteSource *source)
```
Comment (zonque) [inline src/journal-remote/journal-remote-parse.c]: Should be `bool`, not `int`.

File: src/journal-remote/journal-remote-parse.c [hunk 4]
```diff
@@ -464,6 +502,9 @@ int process_source(RemoteSource *source, bool compress, bool seal) {
         assert(source->iovw.iovec);
         assert(source->iovw.count);
 
+        if ( source_monotonic_timestamp_exists(source)==0 )
```
Comment (zonque) [inline src/journal-remote/journal-remote-parse.c]: No spaces after `(`, please.

File: man/systemd.journal-fields.xml (hunk 1 above)
Comment (chaloulo) [inline man/systemd.journal-fields.xml]: Removed blank line

File: src/journal-remote/journal-remote-parse.c (hunk 2 above)
Comment (chaloulo) [inline src/journal-remote/journal-remote-parse.c:460]: Corrected code according to comments.

File: src/journal-remote/journal-remote-parse.c (hunk 3 above)
Comment (chaloulo) [inline src/journal-remote/journal-remote-parse.c]: Changed return type to bool.

File: src/journal-remote/journal-remote-parse.c (hunk 4 above)
Comment (chaloulo) [inline src/journal-remote/journal-remote-parse.c]: Removed spaces after ( and before ).

**Comment by chaloulo** [comment]
Corrected code according to comments/coding style. Also, changed functions source_monotonic_timestamp_exists() and replace_monotonic_timestamp() to be static, as they are not supposed to be exported.

**Comment by chaloulo** [comment]
Pull Request updated:

1) Corrected a logical error: The replacement of the monotonic timestamp should take place even if the _SOURCE_MONOTONIC_TIMESTAMP exists. What should be conditional on the existence of _SOURCE_MONOTONIC_TIMESTAMP is whether the new field _SOURCE_MONOTONIC_TIMESTAMP should be appended.

2) Renamed function replace_monotonic_timestamp() to append_source_monotonic_timestamp().

3) Added OOM check.

**Comment by poettering** [comment]
Not grokking what the rationale here is. We store the boot id for each entry and never compare monotonic times if the boot id of an entry doesn't match. I really don't think that the feature requested here is desirable, or am I missing something?

**Comment by chaloulo** [comment]
In short: 
This PR makes systemd-journal-remote being able to work with multiple sender machines. Without it, journal file on the remote side are continuously rotated, very quickly filling up the disk space.

In detail:
The case that this PR fixes when multiple senders running systemd-journal-upload send logs to a machine running systemd-journal-remote.
Since logs come from multiple machines, it is not guaranteed that they have an increasing monotonic timestamp. This means that it is possible that a log arrives and it has a monotonic timestamp which is smaller than the monotonic timstamp of the previous log, because they originate from different sender machines. 
In this case, this code ([link](https://github.com/systemd/systemd/blob/master/src/journal/journal-file.c#L1517)) will cause journal_file_append_entry() to return -EINVAL. journal_file_append_entry() is called from here ([link](https://github.com/systemd/systemd/blob/master/src/journal-remote/journal-remote-write.c#L146)), so if journal_file_append_entry() returns less than 0, a rotation is made and  journal_file_append_entry() is called again.
The result: If systemd-journal-remote is run with --split-mode=none then all logs are written in the same journal file and we will see continuous rotation of the journal files. In our environment we show thousands of journal files very quickly, and they filled up the disk space.
If --split-mode=host is used, and journal files are separated by the sender hostname the situation is better. But even in this case, if the sender is rebooted, it will cause an unecessary rotation.

Essentially the line that implements the fix is source->ts.monotonic = now(CLOCK_MONOTONIC). 
The functions append_source_monotonic_timestamp() and source_monotonic_timestamp_exists() were created after request by systemd upstream to keep the original timestamp of the log entry in a new field named _SOURCE_MONOTONIC_TIMESTAMP.

Do you think there is a better way to avoid the rotation problem in the case of multiple senders?

**Comment by poettering** [comment]
Oh yuck. Well, but that indicates that the write and rotation logic is borked and needs fixing, not that the entries are patched to work around that to get linear timestamps again.

@keszybz could you look into this, please?

**Comment by chaloulo** [comment]
Before proposing an alternative approach, I would like to discuss about the rationale of what this PR does: It replaces the log's timestamp with the time the log daemon received the log.

It is common that the timestamp of the log entry is the time the logging daemon received the log entry. This is what systemd-journald does also.If the receiving log daemon is systemd-journal-remote it makes sense that the log entry's timestamp is set to the time systemd-journal-remote received the log. To preserve the original timestamp, the _SOURCE_MONOTONIC_TIMESTAMP is added. This is similar to systemd-journald, where the log entry's timestamp is the time systemd-journald received the log, but the field _SOURCE_REALTIME_TIMESTAMP reflects the time that the log entry was generated.

To be consistent with the above thinking, also the ts.realtime field should be replaced, which is not done in the current commit.

Also we need to think about journactl and the time filtering it does with the --since and --until options, when journalctl is used on the remote journal files. Is it better that that the log timestamps are replaced, from the journalctl point of view?

In any case an alternative approach could be to compare the monotonic timestamps in journal_file_append_entry() only if the current_boot_id of the JournalFile is equal to the _BOOT_ID field of the log entry (referring to the if statement at [link](https://github.com/systemd/systemd/blob/master/src/journal/journal-file.c#L1517)).

Do you think the second approach is better? If yes, I can try to do the implementation.

**Comment by poettering** [comment]
Yeah, only enforcing monotonicity if the boot id matches the actual local one sounds like an OK idea. Or instead it would also be easiest if we just drop the entire check... I don't really see what good it does...

**Comment by chaloulo** [comment]
New PR opened (https://github.com/systemd/systemd/pull/2553) with an alternative approach, as suggested by systemd upstream. This PR will be closed.

**Pull request closed without merging**